File: configuration/src/main/java/it/luca/aurora/configuration/metadata/extract/AvroExtract.java
package it.luca.aurora.configuration.metadata.extract;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import it.luca.aurora.configuration.metadata.JsonField;
import org.apache.spark.sql.DataFrameReader;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

import java.util.function.Function;

public class AvroExtract extends Extract {

    @JsonCreator
    public AvroExtract(@JsonProperty(JsonField.TYPE) String type,
                       @JsonProperty(JsonField.FILE_NAME_REGEX) String fileNameRegex) {

        super(type, fileNameRegex);
    }

    @Override
    protected DataFrameReader setUpReader(SparkSession sparkSession) {
        return sparkSession.read().format("avro");
    }

    @Override
    protected Function<String, Dataset<Row>> invokeReader(DataFrameReader reader) {
        return reader::load;
    }
}

File: configuration/src/main/scala/it/luca/aurora/configuration/metadata/extract/AvroExtract.scala
package it.luca.aurora.configuration.metadata.extract

import com.fasterxml.jackson.annotation.JsonProperty
import org.apache.spark.sql.{DataFrame, SparkSession}

/**
 * Coordinates for extracting .avro data
 *
 * @param extractType type of data extraction
 * @param landingPath HDFS path where dataSource data are expected to land
 * @param fileNameRegex regex to be matched by files within landingPath
 */

case class AvroExtract(@JsonProperty(Extract.Type) override val extractType: String,
                       @JsonProperty(Extract.LandingPath) override val landingPath: String,
                       @JsonProperty(Extract.FileNameRegex) override val fileNameRegex: String)
  extends Extract(extractType, landingPath, fileNameRegex) {

  override protected def readDataFrame(sparkSession: SparkSession, path: String): DataFrame =
    sparkSession.read.format("avro").load(path)
}
